### PR TITLE
Sms refactor

### PR DESCRIPTION
**Pull request by judy2k**
Placeholder.

## Contibution Checklist
* [x] Unit tests!
* [x] Updated [CHANGELOG.md]
* [x] My name is in [CONTRIBUTORS.md]

### CODE REVIEW

File: src/main/java/com/nexmo/client/auth/SignatureAuthMethod.java [hunk 1]
```diff
@@ -19,23 +19,31 @@
...
+package com.nexmo.client.auth;
 
-import com.nexmo.client.voice.ncco.TalkNcco;
-import com.nexmo.client.voice.servlet.AbstractAnswerServlet;
-import com.nexmo.client.voice.servlet.NccoResponse;
-import com.nexmo.client.voice.servlet.NccoResponseBuilder;
+import org.apache.http.client.methods.RequestBuilder;
 
-import javax.servlet.http.HttpServletRequest;
+// TODO: This is a stub!
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/auth/SignatureAuthMethod.java]: Should we leave this comment in here?

File: src/main/java/com/nexmo/client/insight/InsightClient.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    public InsightClient(HttpWrapper httpWrapper) {
+        this.basic = new BasicInsightEndpoint(httpWrapper);
+
+    }
+
+    public BasicInsightResponse getBasicNumberInsight(String number) throws IOException, NexmoClientException {
+        return this.basic.execute(new BasicInsightRequest(number));
+    }
+
+    public BasicInsightResponse getBasicNumberInsight(String number, String country) throws IOException,
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: For the next release we might want to add some documentation about what the possible country codes are. Or make `country` an enum.

File: src/main/java/com/nexmo/client/insight/InsightClient.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    public StandardInsightResponse getStandardNumberInsight(String number, String country) throws IOException,
+                                                                                                  NexmoClientException {
+        return this.standard.execute(new StandardInsightRequest(number, country));
+    }
+
+    public AdvancedInsightResponse getAdvancedNumberInsight(String number) throws IOException, NexmoClientException {
+        return this.advanced.execute(new AdvancedInsightRequest(number));
+    }
+
+    public AdvancedInsightResponse getAdvancedNumberInsight(String number, String country, String ipAddress) throws
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: I'm assuming we're skipping Advanced Insight Async for this release?

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    private Integer lookupOutcome;
+    private String lookupOutcomeMessage;
+    private RoamingDetails roaming;
+
+    public static AdvancedInsightResponse fromJson(String json) {
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            return mapper.readValue(json, AdvancedInsightResponse.class);
+        } catch (IOException jpe) {
+            throw new NexmoUnexpectedException("Failed to produce BasicInsightResponse from json.", jpe);
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java]: Failed to produce *AdvancedInsightResponse* from json

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    @JsonProperty("lookup_outcome_message")
+    public String getLookupOutcomeMessage() {
+        return lookupOutcomeMessage;
+    }
+
+    public RoamingDetails getRoaming() {
+        return roaming;
+    }
+
+    public enum PortedStatus {
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java]: Why create this as inner enums?

File: src/main/java/com/nexmo/client/sms/messages/WapPushMessage.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Represents the details of a wap-push message that is to be submitted via the Nexmo REST api.
+ *
+ * @author Paul Cook
+ */
+public class WapPushMessage extends Message {
+    private final String url;
+    private final String title;
+    private Integer validity;
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/sms/messages/WapPushMessage.java]: If we make `VALIDITY` a public enum we could use it here. Maybe next release?

Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/sms/messages/WapPushMessage.java]: Oh nvm, this `validity` is different than the `Validity` returned from a Number Insight. Maybe we should rename one of these or add some documentation to both of them?

File: src/main/java/com/nexmo/client/voice/endpoints/AbstractMethod.java [hunk 7]
```diff
@@ -70,7 +71,9 @@ public AbstractMethod(HttpWrapper httpWrapper) {
     // TODO: Consider wrapping IOException in a nexmo-specific transport exception.
     public ResultT execute(RequestT request) throws IOException, NexmoClientException {
         try {
-            HttpResponse response = this.httpWrapper.getHttpClient().execute(applyAuth(makeRequest(request)).build());
+            HttpUriRequest httpRequest = applyAuth(makeRequest(request)).build();
+            LOG.debug("Request: " + httpRequest);
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/endpoints/AbstractMethod.java]: For future releases we should let the user enable/disable logging

File: src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpointTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
+package com.nexmo.client.insight.advanced;/*
+ * Copyright (c) 2011-2017 Nexmo Inc
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpointTest.java]: Fix copyright position

File: src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightRequestTest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package com.nexmo.client.insight.advanced;/*
+ * Copyright (c) 2011-2017 Nexmo Inc
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightRequestTest.java]: Fix copyright position

File: src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightResponseTest.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
+package com.nexmo.client.insight.advanced;/*
+ * Copyright (c) 2011-2017 Nexmo Inc
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightResponseTest.java]: Fix copyright position

File: src/test/java/com/nexmo/client/sns/SnsEndpointTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
+package com.nexmo.client.sns;/*
+ * Copyright (c) 2011-2017 Nexmo Inc
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/sns/SnsEndpointTest.java]: Fix copyright position

File: src/main/java/com/nexmo/client/auth/SignatureAuthMethod.java (hunk 1 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/auth/SignatureAuthMethod.java]: Nope :)

File: src/main/java/com/nexmo/client/insight/InsightClient.java (hunk 2 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: I'd say there's too many for an enum, but at least we could link to a webpage that lists them in the docs.

File: src/main/java/com/nexmo/client/insight/InsightClient.java (hunk 3 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: Yeah. Too much to think about right now.

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java (hunk 5 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java]: Looking at other APIs, that seems to be the standard way to use them, provided the enum's only used in one place. I'd actually been planning to take some of the top-level enums and move them into their associated classes.

File: src/main/java/com/nexmo/client/sms/messages/WapPushMessage.java (hunk 6 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/sms/messages/WapPushMessage.java]: I might not have it done today, but I plan to document every public method and request and response object, so I'd say at least the latter.

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java (hunk 5 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightResponse.java]: Sounds good

File: src/main/java/com/nexmo/client/voice/endpoints/AbstractMethod.java (hunk 7 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/endpoints/AbstractMethod.java]: They can do that through logging configuration of whichever logging framework they're using behind commons-logging. See what I've done in the Utils class in nexmo-quickstart to configure the default JDK logger.

File: src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpointTest.java (hunk 8 above)
Comment (judy2k) [inline src/test/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpointTest.java]: I can't get my auto-generated test stubs to put this in the right place 😢

**Pull request merged**